README.md
# lyra

Lyra blockchain library for Flutter package project. [Link](https://pub.dev/packages/lyra)

## Getting Started

> dart pub install

## Run Test

> flutter test test/ 

CHANGELOG.md
## 0.0.7

* added a new function: privateKeyIntToPublicKey
## 0.0.6

* added a new function: privateKeyHexToPublicKey
## 0.0.5

* clean dependency

## 0.0.4

* added document to types
  
## 0.0.3

* upgrade dependency
* added example
  
## 0.0.2

* Added github repo url to pub spec.

## 0.0.1

* TODO: Describe initial release.



test/lyra_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:lyra/lyra.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


test/lyra_api_test.dart
import 'package:test/test.dart';
import 'package:lyra/lyra.dart';

void main() {
  test('Get Lyra API', () async {
    const pvk = '2gbESTeBHsgt8um1aNN2dC9jajEDk3CoEupwmN6TRJQckyRbHa';
    //  const pub = 'LUTkgGP9tb4iAFAFXv7i83N4GreEUakWbaDrUbUFnKHpPp46n9KF1QzCtvUwZRBCQz6yqerkWvvGXtCTkz4knzeKRmqid';

    //final api = LyraAPI('devnet', pvk, 'wss://192.168.3.77:4504/api/v1/socket');
    final api = LyraAPI('testnet', pvk, null);
    await api.init();

    var balance = await api.balance();
    print(balance);
    expect(balance, isNotNull);

    if (balance['unreceived']) {
      print('receiving...');
      var balance2 = await api.receive();
      print(balance2);
      expect(balance2, isNotNull);
    }

    // send
    var destAddr =
        'LT8din6wm6SyfnqmmJN7jSnyrQjqAaRmixe2kKtTY4xpDBRtTxBmuHkJU9iMru5yqcNyL3Q21KDvHK45rkUS4f8tkXBBS3';
    var balance3 = await api.send(5.0, destAddr, 'LYR');
    print(balance3);
    expect(balance3, isNotNull);

    api.close();
  });
}


test/lyra_crypto_test.dart
import 'dart:math';

import 'package:test/test.dart';
import 'package:convert/convert.dart';

import 'package:lyra/lyra.dart';

void main() {
  test('ddd', () {
    var hex =
        '3249c9f5976518da19a67048552603e8cf8e54c7b2ee520489656e164134d06d';
    var pub =
        'LUaFA7PZsTPkb6TBfinHYaoGXbecPnLDKtV7vVnkyujnJQgoJytAdfcAH7W3SQETJ4VGKGDxNzNnjNX49WqEH8nPQZ7fA6';
    var pubKey = LyraCrypto.privateKeyHexToPublicKey(hex);
    var accountId = LyraCrypto.lyraEncPub(pubKey);
    print('converted:');
    print(accountId);
    expect(accountId, pub);
  });
  test('key pair generate and verify', () {
    var pvk = '2gbESTeBHsgt8um1aNN2dC9jajEDk3CoEupwmN6TRJQckyRbHa';
    var pub =
        'LUTkgGP9tb4iAFAFXv7i83N4GreEUakWbaDrUbUFnKHpPp46n9KF1QzCtvUwZRBCQz6yqerkWvvGXtCTkz4knzeKRmqid';

    try {
      var ret = LyraCrypto.isAccountIdValid(pub);
      expect(ret, true);

      var ret2 = LyraCrypto.isPrivateKeyValid(pvk);
      expect(ret2, true);

      var pubx = LyraCrypto.prvToPub(pvk);
      expect(pubx, pub);

      var prvKey = LyraCrypto.lyraDecToBytes(pvk);
      var pubKey = LyraCrypto.privateKeyToPublicKey(prvKey);
      var accountId2 = LyraCrypto.lyraEncPub(pubKey);
      expect(accountId2, pub);

      var wallet = LyraCrypto.generateWallet();
      expect(LyraCrypto.isAccountIdValid(wallet[1]), true);
      expect(LyraCrypto.isPrivateKeyValid(wallet[0]), true);

      var msg = 'hello, world!';
      var signtr = LyraCrypto.sign(msg, pvk);
      print(signtr);
      expect(LyraCrypto.verify(msg, pub, signtr), true);
    } catch (e) {
      print(e);
      fail('not verify account id properly: ');
    }
  });
}


lib/lyra.dart
library lyra;

import 'package:json_rpc_2/json_rpc_2.dart';
import 'package:pedantic/pedantic.dart';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'dart:convert';
import 'dart:typed_data';
import 'dart:math';

import 'package:collection/collection.dart';
// ignore: import_of_legacy_library_into_null_safe
import 'package:fast_base58/fast_base58.dart';
import 'package:convert/convert.dart';
// ignore: import_of_legacy_library_into_null_safe
import 'package:pointycastle/ecc/curves/secp256r1.dart';
// ignore: import_of_legacy_library_into_null_safe
import 'package:pointycastle/export.dart';
// ignore: import_of_legacy_library_into_null_safe
import 'package:pointycastle/pointycastle.dart';
// ignore: import_of_legacy_library_into_null_safe
import 'package:pointycastle/signers/ecdsa_signer.dart';

// ignore: import_of_legacy_library_into_null_safe
import 'package:pointycastle/api.dart';
// ignore: import_of_legacy_library_into_null_safe
import 'package:pointycastle/src/impl/secure_random_base.dart';
// ignore: import_of_legacy_library_into_null_safe
import 'package:pointycastle/src/registry/registry.dart';
// ignore: import_of_legacy_library_into_null_safe
import 'package:pointycastle/src/ufixnum.dart';

part 'lyra_api.dart';
part 'lyra_crypto.dart';



lib/lyra_api.dart
part of lyra;

/// API to the Lyra blockchain.
///
/// Lyra provides standard JsonRPC api based on WebSocket from https://mainnet.lyra.live
/// This API is a wrapper of Lyra JsonRPC API.
class LyraAPI {
  String network;
  String? nodeAddress;
  String prvKey;
  String? accountId;
  WebSocketChannel? ws;
  Peer? client;

  /// [network] is 'mainnet', 'testnet', or 'devnet'
  /// [prvKey] is private key.
  /// provides null to [nodeAddress] to use the default url.
  LyraAPI(this.network, this.prvKey, this.nodeAddress);

  /// connect to JsonRPC and get service status.
  Future<void> init() async {
    // crypto

    if (!LyraCrypto.isPrivateKeyValid(prvKey)) {
      throw ('Not valid private key.');
    }
    accountId = LyraCrypto.prvToPub(prvKey);
    nodeAddress ??= 'wss://$network.lyra.live/api/v1/socket';

    // websocket
    var ws = WebSocketChannel.connect(Uri.parse(nodeAddress!));
    client = Peer(ws.cast<String>());

    // The client won't subscribe to the input stream until you call `listen`.
    // The returned Future won't complete until the connection is closed.
    unawaited(client!.listen());

    client!.registerMethod('Notify', (Parameters news) {
      print('Got news from Lyra: ' + news.asList[1].toString());
    });

    client!.registerMethod('Sign', (Parameters req) {
      print('Got signning callback.');

      print('Signing ' + req.asList[1].toString());
      try {
        var signature = LyraCrypto.sign(req.asList[1].toString(), prvKey);
        print('signature is: ' + signature);
        return ['p1393', signature];
      } catch (e) {
        return ['error', e.toString()];
      }
    });

    try {
      var status = await client!.sendRequest('Status', ['2.2.0.0', network]);
      print(status.toString());
    } on RpcException catch (error) {
      print('RPC error ${error.code}: ${error.message}');
    }
  }

  /// send token
  Future<dynamic> send(double amount, String destAddr, String token) async {
    if (client!.isClosed) {
      await init();
    }

    var balanceResp =
        await client!.sendRequest('Send', [accountId, amount, destAddr, token]);
    return balanceResp;
  }

  /// receive token
  Future<dynamic> receive() async {
    if (client!.isClosed) {
      await init();
    }

    var balanceResp = await client!.sendRequest('Receive', [accountId]);
    return balanceResp;
  }

  /// get current balance
  Future<dynamic> balance() async {
    if (client!.isClosed) {
      await init();
    }

    var balanceResp = await client!.sendRequest('Balance', [accountId]);
    return balanceResp;
  }

  /// get transaction history
  Future<dynamic> history(
      DateTime startTimeUtc, DateTime endTimeUtc, int count) async {
    if (client!.isClosed) {
      await init();
    }

    var histResult = await client!.sendRequest('History', [
      accountId,
      startTimeUtc.millisecondsSinceEpoch,
      endTimeUtc.microsecondsSinceEpoch,
      count
    ]);
    return histResult;
  }

  /// close the API, remove private key
  void close() {
    if (client!.isClosed) {
      client!.close();
      accountId = '';
      prvKey = '';
    }
  }
}


lib/lyra_crypto.dart
part of lyra;

/// Help lib for pointycastle library
class NullSecureRandom extends SecureRandomBase {
  static final FactoryConfig factoryConfig =
      StaticFactoryConfig(SecureRandom, 'Null', () => NullSecureRandom());

  var _nextValue = 0;
  @override
  String get algorithmName => 'Null';
  @override
  void seed(CipherParameters params) {}
  @override
  int nextUint8() => clip8(_nextValue++);
}

/// crypto routine for Lyra wallet/blockchain.
class LyraCrypto {
  /// check if private key is valid
  static bool isPrivateKeyValid(String privateKey) {
    try {
      lyraDec(privateKey);
      return true;
    } catch (e) {
      print(e);
      return false;
    }
  }

  /// check if account ID or wallet address is valid
  static bool isAccountIdValid(String accountId) {
    if (accountId.length < 10 || accountId.substring(0, 1) != 'L') return false;
    try {
      lyraDecAccountId(accountId);
      return true;
    } catch (e) {
      print(e);
      return false;
    }
  }

  /// decode account ID to hex encoding of ECDSA public key
  static String lyraDecAccountId(String accountId) {
    var pubKey = accountId.substring(1);

    var decStr = lyraDec(pubKey);
    return '04' + decStr;
  }

  /// decode private key to byte array of ECDSA private key
  static Uint8List lyraDecToBytes(String input) {
    var buff = Base58Decode(input);
    var data = buff.sublist(0, buff.length - 4);
    var crc = checksum(data);
    Function eq = const ListEquality().equals;
    //print("buff len: " + buff.length.toString());
    //print("crc len: " + crc.length.toString());
    //print(buff.sublist(buff.length - 4, buff.length).toString());
    //print(crc.toList());
    if (eq(buff.sublist(buff.length - 4, buff.length), crc)) {
      //print("yes, equal");
      return Uint8List.fromList(data);
    } else {
      //print("no, not equal");
      throw ('Not valid lyra encode string');
    }
  }

  /// decode private key to hex encoding of ECDSA private key
  static String lyraDec(String input) {
    return hex.encode(lyraDecToBytes(input));
  }

  /// Encode with checksum
  static String lyraEnc(List<int> input) {
    var crc = checksum(input);
    return Base58Encode(input + crc);
  }

  /// sha256 digest
  static List<int> sha256(List<int> input) {
    final algo = Digest('SHA-256');
    final hash = algo.process(Uint8List.fromList(input));
    return hash;
  }

  /// crc checksum of byte array
  static List<int> checksum(List<int> input) {
    final h1 = sha256(input);
    var h2 = sha256(h1);
    return h2.sublist(0, 4);
  }

  /// encode a Lyra account ID from ECDSA public key
  static String lyraEncPub(List<int> pubKeyBytes) {
    var ret = lyraEnc(pubKeyBytes.sublist(1));
    return 'L' + ret;
  }

  /// get Account ID from a private key
  static String prvToPub(String prvkey) {
    var prvHex = lyraDec(prvkey);
    var d = BigInt.parse('+' + prvHex, radix: 16);
    var curve = ECCurve_secp256r1();
    var q = curve.G * d;
    var pubKey = ECPublicKey(q, curve);
    var pubKeyBytes = pubKey.Q!.getEncoded(false);

    return lyraEncPub(pubKeyBytes);
  }

  static Uint8List privateKeyToPublicKey(Uint8List prvkey) {
    var prvHex = hex.encode(prvkey);
    return privateKeyHexToPublicKey(prvHex);
  }

  static Uint8List privateKeyHexToPublicKey(String prvHex) {
    var d = BigInt.parse('+' + prvHex, radix: 16);
    return privateKeyIntToPublicKey(d);
  }

  static Uint8List privateKeyIntToPublicKey(BigInt prvInt) {
    var curve = ECCurve_secp256r1();
    var q = curve.G * prvInt;
    var pubKey = ECPublicKey(q, curve);
    var pubKeyBytes = pubKey.Q!.getEncoded(false);

    return pubKeyBytes;
  }

  /// generate a Lyra wallet, return key pair
  static List<String> generateWallet() {
    final rnd = Random.secure();
    var pvkBytes = List<int>.generate(32, (i) => rnd.nextInt(256));
    var pvk = lyraEnc(pvkBytes);
    var pub = prvToPub(pvk);
    return [pvk, pub];
  }

  /// sign a message with private key
  static String sign(String msg, String prvkey) {
    var prvHex = lyraDec(prvkey);
    var d = BigInt.parse('+' + prvHex, radix: 16);

    var eccDomain = ECDomainParameters('secp256r1');
    var privParams = PrivateKeyParameter(ECPrivateKey(d, eccDomain));
    var signParams = ParametersWithRandom(privParams, NullSecureRandom());

    var sig = ECDSASigner(SHA256Digest());
    sig.init(true, signParams);
    var signatur = sig.generateSignature(Uint8List.fromList(utf8.encode(msg)));
    // convert to P1393
    var ecsgn = signatur as ECSignature;
    var rb = hex.decode(ecsgn.r.toRadixString(16));
    var sb = hex.decode(ecsgn.s.toRadixString(16));
    var lst = rb + sb;
    return Base58Encode(lst);
  }

  /// verify a message by signature and account ID
  static bool verify(String msg, String accountId, String signature) {
    var pubHex = lyraDecAccountId(accountId);
    var curve = ECCurve_secp256r1();
    var q = curve.curve.decodePoint(hex.decode(pubHex));

    var eccDomain = ECDomainParameters('secp256r1');
    var pubParams = PublicKeyParameter(ECPublicKey(q, eccDomain));

    var sig = ECDSASigner(SHA256Digest());
    sig.init(false, pubParams);

    // decode P1393
    var lst = Base58Decode(signature);
    var half = lst.length ~/ 2;
    var r = BigInt.parse('+' + hex.encode(lst.sublist(0, half)), radix: 16);
    var s = BigInt.parse('+' + hex.encode(lst.sublist(half, lst.length)),
        radix: 16);

    //print("r: " + r.toString());
    //print("s: " + s.toString());
    var ecsigntr = ECSignature(r, s);
    return sig.verifySignature(Uint8List.fromList(utf8.encode(msg)), ecsigntr);
  }
}


example/example.dart
// @dart=2.9

import 'package:lyra/lyra.dart';

void main(List<String> arguments) async {
  await testApi();
}

/// simple wallet operation test
Future<void> testApi() async {
  const pvk = '2gbESTeBHsgt8um1aNN2dC9jajEDk3CoEupwmN6TRJQckyRbHa';
  //const pub = 'LUTkgGP9tb4iAFAFXv7i83N4GreEUakWbaDrUbUFnKHpPp46n9KF1QzCtvUwZRBCQz6yqerkWvvGXtCTkz4knzeKRmqid';

  //final api = LyraAPI('devnet', pvk, 'wss://192.168.3.77:4504/api/v1/socket');
  final api = LyraAPI('testnet', pvk, null);
  await api.init();

  var balance = await api.balance();
  print(balance);
  //expect(balance, isNotNull);

  if (balance['unreceived']) {
    print('receiving...');
    var balance2 = await api.receive();
    print(balance2);
    //expect(balance2, isNotNull);
  }

  // send
  var destAddr =
      'LT8din6wm6SyfnqmmJN7jSnyrQjqAaRmixe2kKtTY4xpDBRtTxBmuHkJU9iMru5yqcNyL3Q21KDvHK45rkUS4f8tkXBBS3';
  var balance3 = await api.send(5.0, destAddr, 'LYR');
  print(balance3);

  api.close();
}
